VerilogCodingStyle.md

```diff
@@ -1797,92 +1797,171 @@ increment.
 
 ### Don't Cares (`X`'s)
 
-***Explicitly specify don't cares when safe to do so. Don't silently
-squash `X` in logic.***
+***The use of `X` literals in RTL code is strongly discouraged. RTL must not
+assert `X` to indicate "don't care" to synthesis in any case.***
 
-Don't Care values can significantly help the quality of logic optimization and
-should be used wherever it is safe to do so. Explicitly declaring unused decodes
-as `X` has the secondary effect of propagating `X`'s through the design if the
-**input unused** design assumption is violated, making those bugs easier to
-diagnose and fix.
+If not strictly controlled, the use of `X` assignments in RTL to flag invalid or
+don't care conditions can lead to simulation/synthesis mismatches.
 
-Example:
+Instead of assigning and propagating `X` in order to flag and detect invalid
+conditions, it is encouraged to make **extensive use of SVAs**. The added
+benefits of this design practice is that
 
-```systemverilog
-always_comb begin
-  unique case (state)
-    ALPHA:   decode = 16'd1;
-    BETA:    decode = 16'd127;
-    GAMMA:   decode = 16'd43;
-    // all other states are unused:
-    default: decode = 'X;  // or {16{1'bx}} in Verilog-2001
-  endcase
-end
-```
+- No special code style is required to properly propagate `X` conditions,
+- The chance of accidentally introducing simulation/synthesis mismatches is
+  systematically reduced,
+- Simulation breaks early and less signal backtracking is needed to
+  root-cause bugs,
+- In several cases, formal property verification (FPV) can be used to prove
+  whether these SVAs can always be fulfilled.
 
-Write logic that will either propagate `X` or assert when your inputs are `X`.
-Avoid silently squashing `X` by writing logic that resolves an `X` input to a
-non-`X` output. Instead, write logic that either explicitly propagates the `X`
-or uses an assert to raise an exception when an input is `X`.
+The solution presented here has similarities with the approaches presented in
+["Being Assertive With Your X"](http://www.lcdm-eng.com/papers/snug04_assertiveX.pdf)
+by Don Mills.
 
-Be aware: Any logical operation involving `X` always propagates the `X`.
+Note that although don't cares can be used to indicate possible optimization
+opportunities to the synthesis tool, it is debatable whether the gains in logic
+reduction are significant enough to outweigh the possible synthesis mismatch
+issues that the use of `X` literals may entail (especially with the gate-counts
+available in today's technologies). Large LUTs may be an exception to this
+statement and the use of `X` in that context should be carefully evaluated on a
+case-by-case basis.
 
-Example:
+#### Specific Guidance on Case Statements and Ternaries
+
+To comply with this style, RTL must place `` `ASSERT_KNOWN`` assertions on all
```

Comment:
I presume ASSERT_KNOWN is an assertion that a signal isn't X? Have we actually got this assertion defined anywhere? Even if we do we should define what it or point towards a definition in the style guide.

Reply:
Yes, `ASSERT_KNOWN` is essentially what you sometimes refer to as `NEVER_X`. Good catch, I think we only have implicit references to these macros in the style-guide. I.e., if you search for `ASSERT`, you'll find another `ASSERT_INIT` and `ASSERT` reference, but these macros are never explained explicitly. I'll amend this and add a short section that explains these.